### PR TITLE
PYTHON-3052 Add Typings to PyMongo Itself

### PR DESCRIPTION
**Pull request by blink1073**
- [x] Finish manual port of `monitoring.py`, `server_description`, `server_type`, `topology_description`, `url_parser` (all failed when using [`retype`](https://pythonrepo.com/repo/ambv-retype-python-code-refactoring))
- [x] Double check all public interfaces in the auto-ported files:
```
     modified:   pymongo/__init__.py
	modified:   pymongo/collation.py
	modified:   pymongo/common.py
	modified:   pymongo/event_loggers.py
	modified:   pymongo/operations.py
	modified:   pymongo/read_concern.py
	modified:   pymongo/read_preferences.py
	modified:   pymongo/results.py
	modified:   pymongo/write_concern.py
```
- [x] Also look at addressing https://github.com/mongodb-labs/pymongo-stubs/issues/15
  - Used `Union[MutableMapping[str, Any], RawBSONDocumentRef]`
  - Investigate if we can make `Collection` a generic object

### CODE REVIEW

**Comment by blink1073** [comment]
Local experiment with generic document types:

```python
from typing import Any, Mapping, TypeVar, Generic, Optional, Type, Union, cast
from bson.raw_bson import RawBSONDocument
from bson import encode

T = TypeVar('T', bound=Mapping[str, Any])
U = TypeVar('U', bound=Mapping[str, Any])


class Client(Generic[T]):
    document_type: Type[T]

    def __init__(self, document_type: Type[T]) -> None:
        self.document_type = document_type


class CodecOptions(Generic[T]):
    document_type: Type[T]

    def __init__(self, document_type: Type[T]) -> None:
        self.document_type = document_type


class Database(Generic[T]):
    _document_type: Type[T]

    def __init__(self, client: Client[T], codec_options: Optional[CodecOptions[U]] = None) -> None:
        self._document_type = client.document_type

    def bar(self, codec_options: Optional[CodecOptions[U]] = None) -> T:
        return self._document_type()



f = Database(Client(dict))
f.bar()['a'] = 1

g = Database(Client(RawBSONDocument))
# Throws type error: "__setitem__" method not defined on type "RawBSONDocument"
#g.bar()['a'] = 1
```

File: pymongo/collection.py [hunk 1]
```diff
@@ -356,8 +380,13 @@ def with_options(self, codec_options=None, read_preference=None,
                           write_concern or self.write_concern,
                           read_concern or self.read_concern)
 
-    def bulk_write(self, requests, ordered=True,
-                   bypass_document_validation=False, session=None):
+    def bulk_write(
+        self,
+        requests: List[WriteOp],
```
Comment (adavis444) [inline pymongo/collection.py:385]: Given a file
```py
collection = Collection()  # type: ignore
requests = [InsertOne({})]
collection.bulk_write(requests)
```

running mypy gives us
```
error: Argument 1 to "bulk_write" of "Collection" has incompatible type "List[InsertOne]"; expected "List[Union[InsertOne, DeleteOne, DeleteMany, ReplaceOne, UpdateOne, UpdateMany]]"
note: "List" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
note: Consider using "Sequence" instead, which is covariant
```

Changing this line to
```py
        requests: Sequence[WriteOp],
```
resolves this issue.

`common.validate_list("requests", requests)` on [line R455](https://github.com/mongodb/mongo-python-driver/pull/829/files#diff-00c0d0047a08351562fe8a665b8530f6fc909d6634389104a8982e01c190feb5R455) would likely best be replaced by sequence validation instead of list validation.

(This issue also appears in [mongodb-labs/pymongo-stubs](https://github.com/mongodb-labs/pymongo-stubs/blob/d9d1b0d00f54d179ee81174aeb371356776fbe3a/pymongo-stubs/collection.pyi#L67).)

Comment (blink1073) [reply pymongo/collection.py]: Updated to use `Sequence[WriteOp]` and allow tuples as inputs in several places in eb6bf52

File: bson/son.py [hunk 2]
```diff
@@ -136,51 +143,51 @@ def update(self, other=None, **kwargs):
...
                     for k, v in value.items()])
             else:
                 return value
 
-        return transform_value(dict(self))
+        return cast(Dict[_Key, _Value], transform_value(dict(self)))
 
-    def __deepcopy__(self, memo):
-        out = SON()
+    def __deepcopy__(self, memo: Any) -> "SON[Any, Any]":
```
Comment (ShaneHarvey) [inline bson/son.py:186]: I think we can clean this up and revmoe the `cast` like this:
```
    def __deepcopy__(self, memo: MutableMapping[int, "SON[_Key, _Value]"]) -> "SON[_Key, _Value]":

```

File: bson/timestamp.py [hunk 3]
```diff
@@ -54,7 +55,7 @@ def __init__(self, time, inc):
         """
         if isinstance(time, datetime.datetime):
             if time.utcoffset() is not None:
-                time = time - time.utcoffset()
+                time = time - time.utcoffset()  # type: ignore
```
Comment (ShaneHarvey) [inline bson/timestamp.py:58]: Is this ignore a workaround for a mypy bug?
```
bson/timestamp.py:58: error: No overload variant of "__sub__" of "datetime" matches argument type "None"
bson/timestamp.py:58: note: Possible overload variants:
bson/timestamp.py:58: note:     def __sub__(self, timedelta) -> datetime
bson/timestamp.py:58: note:     def __sub__(self, datetime) -> timedelta
bson/timestamp.py:58: note: Right operand is of type "Optional[timedelta]"
```

File: bson/tz_util.py [hunk 4]
```diff
@@ -14,10 +14,12 @@
 
 """Timezone related utilities for BSON."""
 
-from datetime import (timedelta,
+from datetime import (datetime, timedelta,
                       tzinfo)
+from time import time
```
Comment (ShaneHarvey) [inline bson/tz_util.py:19]: "time" is unused AFAICT.

File: gridfs/__init__.py [hunk 5]
```diff
@@ -664,10 +681,10 @@ def upload_from_stream_with_id(self, file_id, filename, source,
         """
         with self.open_upload_stream_with_id(
                 file_id, filename, chunk_size_bytes, metadata,
-                session=session) as gin:
+                session=session) as gin:  # type: GridIn
```
Comment (ShaneHarvey) [inline gridfs/__init__.py:684]: Why is `# type: GridIn` needed here?

File: gridfs/__init__.py [hunk 6]
```diff
@@ -848,6 +865,8 @@ def open_download_stream_by_name(self, filename, revision=-1, session=None):
         query = {"filename": filename}
         _disallow_transactions(session)
         cursor = self._files.find(query, session=session)
+        if revision is None:
+          revision = -1
```
Comment (ShaneHarvey) [inline gridfs/__init__.py:869]: Why this change? Looking at the API we do not support `revision=None`.

File: gridfs/__init__.py [hunk 7]
```diff
@@ -861,8 +880,9 @@ def open_download_stream_by_name(self, filename, revision=-1, session=None):
             raise NoFile(
                 "no version %d for filename %r" % (revision, filename))
 
-    def download_to_stream_by_name(self, filename, destination, revision=-1,
-                                   session=None):
+    def download_to_stream_by_name(self, filename: str, destination: Any,
+                                   revision: Optional[int] = -1,
+                                   session: Optional[ClientSession] = None) -> None:
```
Comment (ShaneHarvey) [inline gridfs/__init__.py:885]: I saw elsewhere you used NoReturn. Should we change all the `-> None:` methods to use it too?

File: gridfs/grid_file.py [hunk 8]
```diff
@@ -572,12 +575,14 @@ def read(self, size=-1):
...
 
-    def readline(self, size=-1):
+    def readline(self, size: Optional[int] = -1) -> bytes:
         """Read one line or up to `size` bytes from the file.
 
         :Parameters:
          - `size` (optional): the maximum number of bytes to read
         """
+        if size is None:
+            size = -1
```
Comment (ShaneHarvey) [inline gridfs/grid_file.py:585]: Why this change? The api does not support size=None and I'd like to avoid adding misc features in this PR.

File: gridfs/grid_file.py [hunk 9]
```diff
@@ -761,15 +769,15 @@ def _next_with_retry(self):
         """
         if self._cursor is None:
             self._create_cursor()
-
+        assert self._cursor is not None
```
Comment (ShaneHarvey) [inline gridfs/grid_file.py:772]: suggest identing this line to avoid duplicating the None check twice in the common case.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
Great work so far! I've reviewed the bson/ and gridfs/ files and think it's probably a good time to stop and address the various comments. Some of them are applicable to all the changes as a whole.

File: bson/binary.py [hunk 10]
```diff
@@ -337,10 +339,10 @@ def __eq__(self, other):
...
 
-    def __hash__(self):
+    def __hash__(self) -> int:
         return super(Binary, self).__hash__() ^ hash(self.__subtype)
 
-    def __ne__(self, other):
+    def __ne__(self, other: Any) -> bool:
         return not self == other
 
     def __repr__(self):
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/binary.py:348]: It is required if we eventually move to strict mode, but I purposely didn't include it for now since it wasn't in the stubs repo

File: bson/__init__.py [hunk 11]
```diff
@@ -61,9 +61,10 @@
 import re
 import struct
 import sys
+from typing import TYPE_CHECKING, Any, BinaryIO, Callable, Dict,  Generator, Iterator, List, Mapping, MutableMapping, NoReturn, Sequence, Tuple, Type, Union, cast
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/__init__.py]: I ran isort on all the files: 0f12b61

File: bson/__init__.py [hunk 12]
```diff
@@ -1027,13 +1036,13 @@ def _decode_all_selective(data, codec_options, fields):
 
     # Decode documents for internal use.
     from bson.raw_bson import RawBSONDocument
-    internal_codec_options = codec_options.with_options(
+    internal_codec_options: CodecOptions = codec_options.with_options(
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/__init__.py:1037]: It had been needed in an earlier iteration, removing it now has no effect

File: bson/codec_options.py [hunk 13]
```diff
@@ -92,6 +98,9 @@ class TypeCodec(TypeEncoder, TypeDecoder):
     pass
 
 
+Codec = Union[TypeEncoder, TypeDecoder, TypeCodec]
+Fallback = Callable[[Any], Any]
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/codec_options.py:100]: Do we want to do that for all types, including the ones in `pymongo.typing`?

File: bson/json_util.py [hunk 14]
```diff
@@ -781,8 +789,8 @@ def default(obj, json_options=DEFAULT_JSON_OPTIONS):
             if not obj.tzinfo:
                 obj = obj.replace(tzinfo=utc)
             if obj >= EPOCH_AWARE:
-                off = obj.tzinfo.utcoffset(obj)
-                if (off.days, off.seconds, off.microseconds) == (0, 0, 0):
+                off = obj.tzinfo.utcoffset(obj)  # type: ignore
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/json_util.py:792]: I added an `assert obj.tzinfo is not None` above and was able to remove the ignore.

File: bson/objectid.py [hunk 15]
```diff
@@ -133,14 +134,14 @@ def from_datetime(cls, generation_time):
             as the generation time for the resulting ObjectId.
         """
         if generation_time.utcoffset() is not None:
-            generation_time = generation_time - generation_time.utcoffset()
+            generation_time = generation_time - generation_time.utcoffset()  # type: ignore
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/objectid.py:135]: Apparently we can't use the `is not None` trick for the result of a function, so I created an intermediate variable and asserted that it was not None.

File: .github/workflows/test-python.yml [hunk 16]
```diff
@@ -26,3 +26,27 @@ jobs:
...
+      - name: Install dependencies
+        run: |
+          python -m pip install -U pip mypy
+          pip install -e ".[zstd, srv]"
+      - name: Run mypy
+        run: |
+          mypy --install-types --non-interactive bson
+          mypy --install-types --non-interactive gridfs
+          mypy --install-types --non-interactive pymongo
+          mypy --install-types --non-interactive test
```
[the remark this answers is not in the document]
Comment (blink1073) [reply .github/workflows/test-python.yml:52]: Done

File: bson/raw_bson.py [hunk 17]
```diff
@@ -119,16 +123,16 @@ class from the standard library so it can be used like a read-only
...
-    def raw(self):
+    def raw(self) -> bytes:
         """The raw BSON bytes composing this document."""
         return self.__raw
 
-    def items(self):
+    def items(self) -> ItemsView[str, Any]:
         """Lazily decode and iterate elements in this document."""
-        return self.__inflated.items()
+        return cast(ItemsView[str, Any], self.__inflated.items())
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/raw_bson.py:130]: Yep, that worked!

File: bson/raw_bson.py [hunk 18]
```diff
@@ -119,16 +123,16 @@ class from the standard library so it can be used like a read-only
...
 
     @property
-    def __inflated(self):
+    def __inflated(self) -> Any:
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/raw_bson.py:133]: Done

File: bson/__init__.py [hunk 19]
```diff
@@ -410,12 +415,12 @@ def _element_to_dict(data, view, position, obj_end, opts):
         return element_name, value, position
 
 
-def _raw_to_dict(data, position, obj_end, opts, result):
+def _raw_to_dict(data: Any, position: int, obj_end: int, opts: Any, result: Any) -> Any:
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/__init__.py:416]: Cool, done

File: bson/__init__.py [hunk 20]
```diff
@@ -88,39 +89,43 @@
...
-BSONCOD = b"\x0D" # Javascript code
-BSONSYM = b"\x0E" # Symbol
-BSONCWS = b"\x0F" # Javascript code with scope
-BSONINT = b"\x10" # 32bit int
-BSONTIM = b"\x11" # Timestamp
-BSONLON = b"\x12" # 64bit int
-BSONDEC = b"\x13" # Decimal128
-BSONMIN = b"\xFF" # Min key
-BSONMAX = b"\x7F" # Max key
+EPOCH_AWARE: datetime.datetime = datetime.datetime.fromtimestamp(0, utc)
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/__init__.py:102]: They are required for strict mode, so I had added them

File: bson/regex.py [hunk 21]
```diff
@@ -49,7 +50,7 @@ class Regex(object):
     _type_marker = 11
 
     @classmethod
-    def from_native(cls, regex):
+    def from_native(cls: Type["Regex"], regex: Pattern[Any]) -> "Regex":
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/regex.py:53]: https://jira.mongodb.org/browse/PYTHON-3062

File: bson/son.py [hunk 22]
```diff
@@ -53,53 +60,53 @@ def __repr__(self):
...
+    def values(self) -> List[_Value]:  # type: ignore[override]
         return [v for _, v in self.items()]
 
-    def clear(self):
+    def clear(self) -> None:
         self.__keys = []
         super(SON, self).clear()
 
-    def setdefault(self, key, default=None):
+    def setdefault(self, key: _Key, default: Optional[Union[_Value, _T]] = None) -> Union[_Value, _T, None]:  # type: ignore[override]
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/son.py:101]: This was copied from https://github.com/mongodb-labs/pymongo-stubs/blob/d23bf60906959f33930fe67f3d9997745a28d6ee/bson-stubs/son.pyi#L29.  I agree it makes sense to remove the `_T` here and in `SON.get`

File: bson/son.py [hunk 23]
```diff
@@ -136,51 +143,51 @@ def update(self, other=None, **kwargs):
...
-    def __eq__(self, other):
+    def __eq__(self, other: Any) -> bool:
         """Comparison to another SON is order-sensitive while comparison to a
         regular dictionary is order-insensitive.
         """
         if isinstance(other, SON):
             return len(self) == len(other) and list(self.items()) == \
                    list(other.items())
-        return self.to_dict() == other
+        return cast(bool, self.to_dict() == other)
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/son.py:158]: Removed

File: bson/son.py [hunk 24]
```diff
@@ -136,51 +143,51 @@ def update(self, other=None, **kwargs):
...
-    def to_dict(self):
+    def to_dict(self) -> Dict[_Key, _Value]:
         """Convert a SON document to a normal Python dictionary instance.
 
         This is trickier than just *dict(...)* because it needs to be
         recursive.
         """
 
-        def transform_value(value):
+        def transform_value(value: Union[List[_Value], Dict[_Key, _Value]]) -> Union[List[_Value], Dict[_Key, _Value]]:
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/son.py:173]: Actually, it turns into a recursive mess of types when I do that.   I just punted and made this private method accept and return `Any`

File: bson/son.py (hunk 2 above)
Comment (blink1073) [reply bson/son.py:185]: Updated

File: bson/timestamp.py (hunk 3 above)
Comment (blink1073) [reply bson/timestamp.py:58]: This is the same issue where the return type of a function can't be used as a "not None" guard

Comment (blink1073) [reply bson/timestamp.py:58]: I added the intermediate variable workaround

File: bson/tz_util.py (hunk 4 above)
Comment (blink1073) [reply bson/tz_util.py:18]: Removed

File: gridfs/__init__.py (hunk 5 above)
Comment (blink1073) [reply gridfs/__init__.py:678]: Not needed anymore, must have been before full typing was added

File: gridfs/__init__.py (hunk 6 above)
Comment (blink1073) [reply gridfs/__init__.py:863]: the `Optional[int]` implies `None` is allowed.  I removed the `Optional`

File: gridfs/__init__.py (hunk 7 above)
Comment (blink1073) [reply gridfs/__init__.py:879]: Those are for functions that always raise an exception

File: gridfs/grid_file.py (hunk 8 above)
Comment (blink1073) [reply gridfs/grid_file.py:582]: This is another case of `Optional[int]`.  In this case the superclass also uses it, we'd have to add an `# ignore: override`:

```
gridfs/grid_file.py:575: error: Argument 1 of "readline" is incompatible with supertype "IOBase"; supertype defines the argument type as "Optional[int]"
gridfs/grid_file.py:575: note: This violates the Liskov substitution principle
gridfs/grid_file.py:575: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
```

File: gridfs/grid_file.py (hunk 9 above)
Comment (blink1073) [reply gridfs/grid_file.py:769]: None

File: bson/timestamp.py [hunk 25]
```diff
@@ -54,7 +55,9 @@ def __init__(self, time, inc):
         """
         if isinstance(time, datetime.datetime):
             if time.utcoffset() is not None:
```
Comment (ShaneHarvey) [inline bson/timestamp.py:57]: Suggest rewriting to remove the duplicate utcoffset() call and redudant assert:
```python
offset = time.utcoffset()
if offset is not None:...
```

File: gridfs/__init__.py [hunk 26]
```diff
@@ -618,14 +628,15 @@ def upload_from_stream(self, filename, source, chunk_size_bytes=None,
            Added ``session`` parameter.
         """
         with self.open_upload_stream(
-                filename, chunk_size_bytes, metadata, session=session) as gin:
+                filename, chunk_size_bytes, metadata, session=session) as gin:  # type: GridIn
```
Comment (ShaneHarvey) [inline gridfs/__init__.py:631]: The type: GridIn comment shouldn't be needed anymore.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: bson/codec_options.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/codec_options.py:100]: Sounds good, updated

File: bson/json_util.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/json_util.py]: I moved the assert below that line for clarity

File: gridfs/grid_file.py (hunk 8 above)
Comment (blink1073) [reply gridfs/grid_file.py:582]: Agreed, but because `io.IOBase` uses `Optional[int]`, we have to add an ignore override comment.  I'll do that.

File: gridfs/grid_file.py [hunk 27]
```diff
@@ -637,18 +639,19 @@ def seek(self, pos, whence=_SEEK_SET):
...
         if new_pos == self.__position:
-            return
+            return new_pos
 
         self.__position = new_pos
         self.__buffer = EMPTY
         if self.__chunk_iter:
             self.__chunk_iter.close()
             self.__chunk_iter = None
+        return new_pos
```
[the remark this answers is not in the document]
Comment (blink1073) [reply gridfs/grid_file.py:649]: Done

File: bson/objectid.py [hunk 28]
```diff
@@ -133,14 +132,16 @@ def from_datetime(cls, generation_time):
             as the generation time for the resulting ObjectId.
         """
         if generation_time.utcoffset() is not None:
-            generation_time = generation_time - generation_time.utcoffset()
+            offset =  generation_time.utcoffset()
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/objectid.py:135]: Agreed, updated

File: bson/son.py [hunk 29]
```diff
@@ -20,29 +20,36 @@
 
 import copy
 import re
-
 from collections.abc import Mapping as _Mapping
+from typing import (Any, Dict, Iterable, Iterator, List, Mapping,
+                    MutableMapping, Optional, Pattern, Tuple, Type, TypeVar,
+                    Union, cast)
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/son.py:26]: Another thing we can catch with `pre-commit`.  ;)

File: bson/son.py [hunk 30]
```diff
@@ -176,14 +183,16 @@ def transform_value(value):
...
-    def __deepcopy__(self, memo):
-        out = SON()
+    def __deepcopy__(self, memo: MutableMapping[int, "SON[_Key, _Value]"]) -> "SON[_Key, _Value]":
+        out: SON[_Key, _Value] = SON()
         val_id = id(self)
         if val_id in memo:
-            return memo.get(val_id)
+            value = memo.get(val_id)
+            assert value is not None
+            return value
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/son.py:192]: Ideally `if val_id in memo` would guard against the `get()` being `None`, but it does not, which is why I added the intermediate variable and guard.

File: bson/son.py [hunk 31]
```diff
@@ -176,14 +183,16 @@ def transform_value(value):
...
         memo[val_id] = out
         for k, v in self.items():
             if not isinstance(v, RE_TYPE):
-                v = copy.deepcopy(v, memo)
+                v = copy.deepcopy(v, dict(memo))
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/son.py:196]: Done

File: bson/son.py [hunk 32]
```diff
@@ -136,35 +143,35 @@ def update(self, other=None, **kwargs):
...
-    def __eq__(self, other):
+    def __eq__(self, other: Any) -> bool:
         """Comparison to another SON is order-sensitive while comparison to a
         regular dictionary is order-insensitive.
         """
         if isinstance(other, SON):
             return len(self) == len(other) and list(self.items()) == \
                    list(other.items())
-        return self.to_dict() == other
+        return  self.to_dict() == other
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/son.py:159]: Fixed

File: bson/timestamp.py (hunk 25 above)
Comment (blink1073) [reply bson/timestamp.py:57]: Done

File: gridfs/__init__.py (hunk 26 above)
Comment (blink1073) [reply gridfs/__init__.py:631]: Confirmed and removed

**Comment by blink1073** [comment]
Note: I removed the changes to `pymongo` and `test` in order to make this PR easier to review.  I will follow up with those in other PRs.

File: gridfs/grid_file.py (hunk 8 above)
Comment (ShaneHarvey) [reply gridfs/grid_file.py]: Oh I see. Yeah override is what typeshed does: https://github.com/python/typeshed/blob/df0fe1645681638359a4ab706732e79c62db6a71/stdlib/io.pyi#L128

File: test/mockupdb/operations.py [hunk 33]
```diff
@@ -14,15 +14,14 @@
 
 from collections import namedtuple
 
-from mockupdb import *
-from mockupdb import OpMsgReply
+from mockupdb import OpMsgReply, OpMsg, OpReply  # type: ignore
```
Comment (ShaneHarvey) [inline test/mockupdb/operations.py:17]: What if we use the mypy.ini file to ignore all mockupdb imports instead:
```
[mypy-mockupdb]
ignore_missing_imports = True
```

File: test/performance/perf_test.py [hunk 34]
```diff
@@ -44,7 +45,7 @@
 
 OUTPUT_FILE = os.environ.get('OUTPUT_FILE')
 
-result_data = []
+result_data: List[Any] = []
```
Comment (ShaneHarvey) [inline test/performance/perf_test.py:48]: Isn't this annotation redundant?

File: bson/son.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply bson/son.py:191]: Yeah I agree but I think it's similar to the other issue since mypy can't determine if the val_id might be removed by another thread in between the if-statement and the get() call. Regardless, I think we can replace these 3 lines with `return memo[val_id]` since we know val_id is always there. Then we don't need to add redundant code just to appease mypy.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
Thanks for pushing out the pymongo/+test/ changes!

File: test/performance/perf_test.py (hunk 34 above)
Comment (blink1073) [reply test/performance/perf_test.py:48]: `test/performance/perf_test.py:48: error: Need type annotation for "result_data" (hint: "result_data: List[<type>] = ...")`

File: doc/changelog.rst [hunk 35]
```diff
@@ -21,6 +21,8 @@ PyMongo 4.0 brings a number of improvements including:
   $merge and $out executing on secondaries on MongoDB >=5.0.
   aggregate() now always obeys the collection's :attr:`read_preference` on
   MongoDB >= 5.0.
+- :meth:~gridfs.GridOut.seek` now returns the new position in the file, to
```
[the remark this answers is not in the document]
Comment (blink1073) [reply doc/changelog.rst:24]: Fixed

File: test/mockupdb/operations.py (hunk 33 above)
Comment (blink1073) [reply test/mockupdb/operations.py:17]: Done

File: bson/son.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (blink1073) [reply bson/son.py:191]: Used type ignore instead

File: test/performance/perf_test.py (hunk 34 above)
Comment (ShaneHarvey) [reply test/performance/perf_test.py]: Oh that's surprising. I thought it would default to List. TIL.

File: bson/son.py [hunk 36]
```diff
@@ -176,11 +182,11 @@ def transform_value(value):
...
         return transform_value(dict(self))
 
-    def __deepcopy__(self, memo):
-        out = SON()
+    def __deepcopy__(self, memo: Dict[int, "SON[_Key, _Value]"]) -> "SON[_Key, _Value]":
+        out: SON[_Key, _Value] = SON()
         val_id = id(self)
         if val_id in memo:
-            return memo.get(val_id)
+            return memo.get(val_id)  # type: ignore
```
Comment (ShaneHarvey) [inline bson/son.py:189]: We can make this type-safe and remove `# type: ignore` with `return memo[val_id]` instead of using .get().

File: test/performance/perf_test.py (hunk 34 above)
Comment (blink1073) [reply test/performance/perf_test.py]: It defaulted to `List[unknown]`, but adding `:  List` allowed it to reference as `List[Any]`, so the `[Any]` was redundant.

File: bson/son.py (hunk 36 above)
Comment (blink1073) [reply bson/son.py]: Oh, ha, duh, thanks

**Review by ShaneHarvey — APPROVED** [review]
LGTM!
